docs/oid-info.md

```diff
@@ -182,16 +182,16 @@ that Sigstore operates.
 
 | GitHub [(docs)][github-oidc-doc] | GitLab | CircleCI | Buildkite | Fulcio Certificate Extension | Why / Notes / Questions                                                                                                                                                                |
 |--------------------|--------|----------|-----------|------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
-| aud                | ??     | aud       | aud        | N/A                          | Only used to validate the JWT.                                                                                                                                                         |
+| aud                | aud     | aud       | aud        | N/A                          | Only used to validate the JWT.                                                                                                                                                         |
 | iss                | iss     | iss       | iss        | Issuer                       | This already exists. For example: https://token.actions.githubusercontent.com                                                                                                          |
 | exp                | exp     | exp       | exp        | N/A                          | Only used to validate the JWT.                                                                                                                                                         |
 | nbf                | nbf     | nbf       | nbf        | N/A                          | Only used to validate the JWT.                                                                                                                                                         |
 | iat                | iat     | iat       | iat        | N/A                          | Only used to validate the JWT.                                                                                                                                                         |
-| server_url + job_workflow_ref   | ??     | ??       | ??        | Build Signer URI             | Reference to specific build instructions that are responsible for signing. Can be the same as Build Config URI. For example a reusable workflow in GitHub Actions or a Circle CI Orbs. |
+| server_url + job_workflow_ref   | server_url + project_path + /-/blob/ + ci_commit_sha + / + ci_config_path | ??       | ??        | Build Signer URI             | Reference to specific build instructions that are responsible for signing. Can be the same as Build Config URI. For example a reusable workflow in GitHub Actions or a Circle CI Orbs. |
```

Comment:
Example URL: https://gitlab.com/gitlab-org/gitlab/-/blob/d33e88e2d80b97b0c15e5c11aaf90e17d0916fda/.gitlab-ci.yml

Should this be the same as Build Config URI?

Reply:
> Additionally, include is not versioned and it could change between multiple different CI pipeline runs.

@Brcrwilliams the immutability problem is not unique to us and it can't be solved in this context anyway. A workflow ref in GitHub Actions has no immutability guarantee either. Folks should be version pinning their `include` references by SHA if they want such guarantees. For local `include` references (like in your example) it doesn't even matter because the ref is the same as the root `.gitlab-ci.yml` pipeline definition.

We do need to account for [custom CI/CD config file locations](https://docs.gitlab.com/ee/ci/pipelines/settings.html#specify-a-custom-cicd-configuration-file) which can be set to anything including arbitrary remote HTTP endpoints. As such we may need to normalize this claim value to a fully qualified URI in all cases for consistency. We may need to manually checksum references to files that are not stored in git.

cc @wlynch

Reply:
I think we should name the new claims `pipeline_ref` and `pipeline_sha`. For the reasons noted above, `ci_config_path` is not always going to be `.gitlab-ci.yml` and the checksum is not necessarily based on `ci_commit_sha`.

Also note, we will be making some pretty significant changes to pipeline definitions in the future (https://gitlab.com/gitlab-org/gitlab/-/issues/398129#note_1323114974) and may wish to change the URI format of `pipeline_ref` to align with the syntax of [CI pipeline components](https://docs.gitlab.com/ee/architecture/blueprints/ci_pipeline_components/). Keeping that logic for constructing that on our side is another good reason to include fully qualified URIs in the claim.

Reply:
@wlynch just for clarification, other providers do not appear to provide digests for the fully expanded workflow/pipeline/job definition either. GitHub, for example, is only providing the commit SHA in their `workflow_sha` claim.

I agree this will be great to have eventually, but unless I'm missing something, including the commit SHA in "Build Config Digest" would be consistent with GitHub.

Reply:
GitHub is a bit different in that it makes the distinction between `workflow_sha` for the entire workflow vs `job_workflow_sha` that's just for the particular job.

We don't have a similar concept of referencing a digest for a job in GitLab. While we can use the pipeline_sha to try to and get the job, I don't think it actually gives us much extra value than if we omitted it.

I was thinking that it would look something like:

Field | Value
-- | --
Build Signer URI | https://gitlab.com/org/repo/-/job/1234
Build Signer Digest | 
Build Config URI | https://gitlab.com/org/repo/.gitlab-ci.yml (pipeline_ref)
Build Config SHA | abcdef1234 (pipeline_sha)
Run Invocation URI | https://gitlab.com/org/repo/-/pipeline/1234

Reply:
@wlynch ah, agreed! I missed that "Build Signer Digest" was a separate field. We will revisit once CI components are implemented: https://docs.gitlab.com/ee/architecture/blueprints/ci_pipeline_components/#definition-of-pipeline-component

Reply:
@haydentherapper - [it was pointed out](https://gitlab.com/gitlab-org/gitlab/-/issues/388517#note_1350509539) that https://github.com/sigstore/fulcio/blob/main/docs/oid-info.md says:

> - MUST include claim to support: `Build Signer Digest` which is an immutable reference to a specific version of the build instructions that are responsible for signing.

This is a case where we don't have a good value to put here right now, though I think there's general agreement to work towards this.

Off the top of my head, I think other CI providers like Tekton and Cloud Build that let you inline build instructions would also be impacted because there isn't a specific digest to source from (though you can still fetch the build instructions via the URI).

While I agree this is important for verifying integrity, this feels like we may have over-indexed on GitHub Actions and made the barrier a little too high to start.
I'm thinking we should drop this down to `SHOULD` - wdyt?

Reply:
My concern would be that the Build Signer URI is mutable, effectively making the identity mutable, so there's no way to link back to the actual build instructions. With that said, if none of the other CI platforms provide this, we might have to drop MUST.

cc @laurentsimon @asraa @feelepxyz if you have more thoughts

Reply:
The Build Signer URI (in GitLab's case our plan is to use Job ID) would still be useful for asserting identity since it is considered a unique immutable identifier for the service - though it places trust on GitLab to do the right thing and limits a client's ability to detect if the CI service misbehaves.

I agree that having a digest that can be verified independently by clients is something that CI providers *should* do - but I think having this requirement as `MUST` is a high bar that not all CI providers may be able to do right away.

Reply:
Im not opposed to changing it to SHOULD, but I’d like to get some more input from GH, given we have discussed this as a requirement.

Reply:
I think we are fine to lower this to SHOULD. Even for GitHub supporting the `Build Signer Digest` is a recent addition. By lowering this to SHOULD we still allow policy makers to REQUIRE this value, which I think is a good tradeoff.

Reply:
@wlynch Can you create a separate PR to update this?

Reply:
Of course! Will do.